### PR TITLE
[OLED-259] Comply Shard stealing with streamMaxShards

### PR DESCRIPTION
**Pull request by phoebe-canva**
### Context
* Added streamMaxShards config in this PR - https://github.com/Canva/benthos/pull/1
* However the stealing still occurs and doesnt comply with the configuration because that part does not have the additional check

### Changes made
* Moved the part of the function into the function called `stealShards` so it is testable and added the `streamMaxShards` check
* Added tests to ensure it is working as expected
* Fixed up the logger message

### CODE REVIEW

File: internal/impl/aws/input_kinesis.go [hunk 1]
```diff
@@ -692,6 +648,8 @@ func (k *kinesisReader) runExplicitShards() {
 	}
 }
 
+// This function attempts to try and claim shards only if the number of shards already claimed is less than the max number of shards
```
Comment (sasanrose) [inline internal/impl/aws/input_kinesis.go]:
```suggestion
// maxShardRunBalancedConsumer attempts to try and claim shards only if the number of shards already claimed is less than the max number of shards
```

Convention is to start the doc blocks with the name of the function/struct you are leaving the comment for

File: internal/impl/aws/input_kinesis.go [hunk 2]
```diff
@@ -722,6 +680,67 @@ func (k *kinesisReader) maxShardRunBalancedConsumer(wg *sync.WaitGroup, streamID
 	return numShardClaimed, nil
 }
 
+// Function that attempts to steal shards if it meets the following condition:
```
Comment (sasanrose) [inline internal/impl/aws/input_kinesis.go]:
```suggestion
// stealShards attempts to steal shards if it meets the following condition:
```

File: internal/impl/aws/input_kinesis_test.go [hunk 3]
```diff
@@ -16,6 +21,8 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
+// These mocks are used to mock the underlying AWS Api that is called by runConsumer()
```
Comment (sasanrose) [inline internal/impl/aws/input_kinesis_test.go]:
```suggestion
// mockDynamoDbForKinesis is used to mock the underlying AWS Api that is called by runConsumer()
```

File: internal/impl/aws/input_kinesis.go [hunk 4]
```diff
@@ -593,53 +593,9 @@ func (k *kinesisReader) runBalancedShards() {
...
-					if err = k.runConsumer(&wg, streamID, randomShard, sequence); err != nil {
-						k.log.Errorf("Failed to start consumer: %v\n", err)
-					} else {
-						// If we successfully stole the shard then that's enough
-						// for now.
-						break
-					}
-				}
+			// Note: Stealing only happens if all shards have already been claimed
+			if _, err = k.stealShards(&wg, streamID, clientClaims); err!=nil {
```
Comment (sasanrose) [inline internal/impl/aws/input_kinesis.go]: I think you need to to format this file using goimports

File: internal/impl/aws/input_kinesis_test.go [hunk 5]
```diff
@@ -30,30 +37,56 @@ func (m *mockDynamoDbForKinesis) GetItemWithContext(aws.Context, *dynamodb.GetIt
...
+		Config:          session.NewConfig(),
+		Streams:         streams,
+		StreamsMaxShard: streamsMaxShard,
+		DynamoDB:        input.NewDynamoDBCheckpointConfig(),
+		CheckpointLimit: 1024,
+		CommitPeriod:    "5s",
+		LeasePeriod:     "30s",
+		RebalancePeriod: "30s",
+		StartFromOldest: true,
+		Batching:        batchconfig.NewConfig(),
```
Comment (sasanrose) [inline internal/impl/aws/input_kinesis_test.go:75]: nit: Do we use these fields in our tests? Recommend to remove them if not required to avoid cluttering the test

File: internal/impl/aws/input_kinesis.go [hunk 6]
```diff
@@ -722,6 +680,67 @@ func (k *kinesisReader) maxShardRunBalancedConsumer(wg *sync.WaitGroup, streamID
 	return numShardClaimed, nil
 }
 
+// Function that attempts to steal shards if it meets the following condition:
+// * It is not its own clientId
+// * It has not reached the max shard number this client can claim
+// * The client it is stealing from has 2 or more shards more than this client currently has
+// It returns True if successful in stealing a shard and any errors if it encountered any
+func (k *kinesisReader) stealShards(wg *sync.WaitGroup, streamID string, clientClaims map[string][]awsKinesisClientClaim) (bool, error){
```
Comment (sasanrose) [inline internal/impl/aws/input_kinesis.go]: nit: rename this to stealShard as it seems it only steals one shard

File: internal/impl/aws/input_kinesis.go [hunk 7]
```diff
@@ -722,6 +680,67 @@ func (k *kinesisReader) maxShardRunBalancedConsumer(wg *sync.WaitGroup, streamID
...
+	// There were no unclaimed shards, let's look for a shard to steal.
+	selfClaims := len(clientClaims[k.clientID])
+	maxShardNumber, maxShardExists := k.streamMaxShards[streamID]
+	isSuccess:= false
```
Comment (sasanrose) [inline internal/impl/aws/input_kinesis.go]: Maybe it's easier to reuse the `ErrLeaseNotAcquired` error rather than having a `isSuccess` which is only used for test purpose.
So the function will return only an error and then:
- when `isSuccess` is true (we managed to steal a shard) instead of `break` we can `return nil`
- and after the `for` leap we can return `ErrLeaseNotAcquired`
- then the caller can check if the error returned by the function is `ErrLeaseNotAcquired` then it can ignore the error
- We can also update tests to check the returned error type rather than check `isSuccess`

File: internal/impl/aws/input_kinesis.go [hunk 8]
```diff
@@ -692,6 +648,8 @@ func (k *kinesisReader) runExplicitShards() {
 	}
 }
 
+// This function attempts to try and claim shards only if the number of shards already claimed is less than the max number of shards
+// It returns the number of shards it has successfully claimed and any errors it encountered
```
Comment (sasanrose) [inline internal/impl/aws/input_kinesis.go:655]: nit: I know not part of this PR but I think when an error happen in this function here:
```
sequence, err := k.checkpointer.Claim(k.ctx, streamID, shardID, clientID)
			if err != nil {
				if k.ctx.Err() != nil {
					return 0, err
				}
```
We should still return the number of so far claimed shards rather than 0

**Review by sasanrose — CHANGES_REQUESTED** [review]
Great work Phoebe, left a couple of comments

File: internal/impl/aws/input_kinesis.go (hunk 1 above)
Comment (phoebe-canva) [reply internal/impl/aws/input_kinesis.go]: Gotcha thanks!

File: internal/impl/aws/input_kinesis_test.go (hunk 5 above)
Comment (phoebe-canva) [reply internal/impl/aws/input_kinesis_test.go:75]: I have to pass in something to these data fields or it will complain 😓

File: internal/impl/aws/input_kinesis.go (hunk 7 above)
Comment (phoebe-canva) [reply internal/impl/aws/input_kinesis.go]: Ah okay, thanks :) That would make it slightly cleaner, i was thinking that it is not _technically_ an error so it might add additional confusion. I will put a comment down about it :) I will create a new error because `ErrLeaseNotAcquired` is a very specific error 😅

File: internal/impl/aws/input_kinesis.go (hunk 8 above)
Comment (phoebe-canva) [reply internal/impl/aws/input_kinesis.go:655]: oh true, its fixed now :)

File: internal/impl/aws/input_kinesis.go (hunk 4 above)
Comment (phoebe-canva) [reply internal/impl/aws/input_kinesis.go]: gotcha 👍

File: internal/impl/aws/input_kinesis.go [hunk 9]
```diff
@@ -593,53 +594,10 @@ func (k *kinesisReader) runBalancedShards() {
...
-						k.log.Errorf("Failed to start consumer: %v\n", err)
-					} else {
-						// If we successfully stole the shard then that's enough
-						// for now.
-						break
-					}
-				}
+			// Note: Stealing only happens if all shards have already been claimed
+			err = k.stealShard(&wg, streamID, clientClaims)
+			if !(err == nil || err == errShardNotStolen) {
```
Comment (sasanrose) [inline internal/impl/aws/input_kinesis.go]: Could we please use `errors.Is(err, errShardNotStolen)` instead?

Also we still need to handle cases where error is not `errShardNotStolen`. For instance for `return k.ctx.Err()`

It's a bit confusing though as in both cases we should just return?!

File: internal/impl/aws/input_kinesis_test.go [hunk 10]
```diff
@@ -148,3 +182,103 @@ func TestRunMaxShardConfigBalancedShards(t *testing.T) {
...
+
+	testCases := []struct {
+		testName          string
+		streams           []string
+		streamsMaxShard   []string
+		clientClaims      map[string][]awsKinesisClientClaim
+		streamNameToClaim string
+		clientId          string
+		expectedIsError   bool
+		errShardNotStolen bool
```
Comment (sasanrose) [inline internal/impl/aws/input_kinesis_test.go:204]: nit: Could we combine this two fields to `expectedError` of type `error`. Then in test cases it can either be `nil` or `errShardNotStolen`

File: internal/impl/aws/input_kinesis.go [hunk 11]
```diff
@@ -722,6 +682,65 @@ func (k *kinesisReader) maxShardRunBalancedConsumer(wg *sync.WaitGroup, streamID
 	return numShardClaimed, nil
 }
 
+// stealShard() attempts to steal shards if it meets the following condition:
+// * It is not its own clientId
+// * It has not reached the max shard number this client can claim
+// * The client it is stealing from has 2 or more shards more than this client currently has
+// It returns nil if it succesfully steals a shard, otherwise returns error errShardNotStolen if shards are not stolen or any errors if it encountered any
+func (k *kinesisReader) stealShard(wg *sync.WaitGroup, streamID string, clientClaims map[string][]awsKinesisClientClaim) ( error) {
```
Comment (sasanrose) [inline internal/impl/aws/input_kinesis.go]:
```suggestion
func (k *kinesisReader) stealShard(wg *sync.WaitGroup, streamID string, clientClaims map[string][]awsKinesisClientClaim) error {
```

I'm sorry but formatting is pretty serious in Go, lol. I'm pretty sure our PR will be rejected by Benthos if formatting is not right that's why I'm picky. I think you need to setup your go formatter in your IDE.

Comment (phoebe-canva) [reply internal/impl/aws/input_kinesis.go]: hmm i already have setup my goformatter to use goimport though

File: internal/impl/aws/input_kinesis_test.go (hunk 10 above)
Comment (phoebe-canva) [reply internal/impl/aws/input_kinesis_test.go:204]: i know in my test cases that both are the same - but in the future there will be scenarios where we do expect an error but it is NOT `errShardNotStolen`

Comment (phoebe-canva) [reply internal/impl/aws/input_kinesis_test.go:204]: The problem imho stems from the change that we made making `errShardNotStolen` an error, which _technically_ it is not

File: internal/impl/aws/input_kinesis.go (hunk 9 above)
Comment (phoebe-canva) [reply internal/impl/aws/input_kinesis.go]: > It's a bit confusing though as in both cases we should just return?!

it does work though :) if there is no error OR  it is shard not stolen == true and then with the negation, it would be `false`
whereas if there is an error, it would be false and with the negation it would be true.

Anyways, i changed the conditional around, hopefully easier to read

File: internal/impl/aws/input_kinesis.go (hunk 11 above)
Comment (sasanrose) [reply internal/impl/aws/input_kinesis.go]: hmm, strange. It may be not setup to run on save?!

Comment (phoebe-canva) [reply internal/impl/aws/input_kinesis.go]: ahhh i think it was silently failing on vscode :( i fixed it now, hopefully no more 😅

**Review by sasanrose — APPROVED** [review]
LGTM!
